### ISSUE TITLE
Azure Topic Receving - Acknowledgement

### ISSUE THREAD

**Issue by Romanx** [noodlefrenzy/node-amqp10#101]
Hey,

I've been playing with this library using azure service bus topics and haven't been able to find a way to acknowledge messages once I've processed them to avoid loosing messages.

Is there currently any method in place to do this? The Azure node library has *peekLock* which gets the message and locks it until you use *deleteMessage* yourself.

Thanks,

**Comment by noodlefrenzy** [comment]
In the "Caveats and Todos" I mention:

> Disposition frames are not dealt with properly, and thus message lifecycles aren't tracked correctly. Specifically, we don't send disposition frames on receipt, and we don't send proper "unsettled" information when re-attaching links.

Currently the policy is set to auto-settle by default (so you shouldn't see repeats, but can't "peek-lock"). AMQP has no built-in "peek-lock" but is instead a combination of not auto-settling, and sending disposition frames on completion. I've added code to allow sending of disposition frames (primarily to try and cope with ActiveMQ's bad behavior), but haven't surfaced that yet. A V2 of the library is coming _soon_ (there are some reconnect issues we're sorting through) and I'll make sure it makes it into that.

**TL;DR version**: Not currently. Hopefully soon.

I'll keep this open until that ships.

**Comment by Romanx** [comment]
Hey, I read that but didn't entirely understand that it related to this. Thank you for clarifying and I'll look forward to the progress :)

**Comment by mbroadst** [comment]
@Romanx I'm currently merging in more fine-grained control for message disposition. Currently merged in to master is support for automatic and manual disposition.  You can see some examples (here)[https://github.com/noodlefrenzy/node-amqp10/blob/master/test/integration/qpid/disposition.test.js]. Currently the manual disposition (reject, accept, etc) are a little awkward, but that will be soon resolved with the ReceiverLink, SenderLink split coming shortly.

**Comment by Romanx** [comment]
Hey,

Thanks,for getting this done so quickly. I'm currently working on a side project so i'll likely look at this next when the ReceiverLink and SenderLink are implemented.

**Comment by mbroadst** [comment]
@Romanx ReceiverLink and SenderLinks have been implemented in master, it would be great to get some feedback on what kind of API you would want to see there (as we can't encompass all use cases :smile:).

**Comment by mbroadst** [comment]
@Romanx for an example of manual disposition you can take a look at the disposition.test.js in the QPID integration tests

**Comment by mbroadst** [comment]
closing until more data comes in, this should be fixed in the 2.0.x releases

**Comment by Romanx** [comment]
Thanks for the great work @mbroadst and @noodlefrenzy. Sorry for the lack of reply, haven't had a chance to get back to my side project since!

**Comment by Romanx** [comment]
Hi there @mbroadst,

I had a chance to update my side project to the 2.0.0 code and all seemed to be going well (once I found the silly typos I made). The only thing is that it wasn't exactly acting as I expected in certain cases.

I added manual disposition as per the test you recommended above. When I deliberately threw an error I was expecting the result to be similar to the standard Azure node.js lib with `ack` and `nak`. As i didn't acknowledge the result i was expected it to be released for another reciever. It seems however that it was `nak'd` and marked as failed in my service bus.

Am I doing something wrong or is this expected behaviour? I'd think that if i didn't explicitly accept or reject the message it shouldn't be altered at all. This way if there's something that kills my consumer (for whatever reason) the message can be attempted by another or the newly resurrected consumer.

I'd be interested in your thoughts

**Comment by mbroadst** [comment]
@Romanx if you are set up for manual disposition then you _must_ manually settle the incoming messages in all cases. I'm not sure what you're talking about with "deliberately threw an error", do you mean you did a `throw new Error` from within your receive callback? We could theoretically add support for that (automatically reject the message if an error was thrown), but for now its expected behavior that a message is not settled if you haven't explicitly done so (in manual disposition mode).

Did that answer the question? Perhaps a code example from you could illuminate the use case for us

**Comment by noodlefrenzy** [comment]
@mbroadst My worry is @Romanx is saying that he error'd out (deliberately) prior to settling the message in either direction, and instead of SB timing the message out and letting it be re-delivered, it instead just threw up its hands and marked it as dead. In theory, SB is supposed to time-bomb the message and re-deliver if it's not ack/nak'd in time.

@Romanx am I correct in what you were trying to do? I'm trying to write an integration test to see if I can flush out this case, so I'd also love a code sample if you have one :)

**Comment by Romanx** [comment]
Hi,

Yes that's exactly what I did @noodlefrenzy , I threw an error in the receive callback. It might be a nice user experience if it performed an action in that case (as my instinct was that it would do something). What would happen in the case that I didn't settle an incoming message?

```javascript
receiver.on('message', function(message) { 
    //Do Stuff
    
    if(error) throw new Error('error');    
    
    this.accept(message);
});
```
In this case i'd like to accept only if everything succeeded until the last statement, it makes sense that if there's an error case (like above) that I perform `this.reject(message)` however what about the case where something causes my consumer to crash and i'd like the message to attempt redelivery to another consumer?

**Comment by mbroadst** [comment]
@Romanx yeah okay that's what I thought you meant. @noodlefrenzy we can pop a try/catch around the call to the callback on the `ReceiverLink` and take some sort of default action (`reject` seems a likely candidate). I agree that this is more idiomatic. I'm not sure, however, whether this should be an opt-in feature or not? thoughts?

**Comment by Romanx** [comment]
@mbroadst with a reject a message never attempts redelivery correct? I'm not sure that's a nice experience as it could be a simple error with a child service (a db connection maybe) which causes your error, not the message itself. This way that message never attempts redelivery when it would succeed.

**Comment by mbroadst** [comment]
@Romanx I'll defer to @noodlefrenzy and you for what Azure's behavior is, but from my reading of the spec it seems to indicate that it will simply reduce the deliveryCount, nothing indicating that it would not attempt to be redelivered:

```
At the target, the rejected outcome is used to indicate that an incoming message is invalid and therefore unprocessable.
The rejected outcome when applied to a message will cause the delivery-count to be incremented in
the header of the rejected message.
At the source, the rejected outcome means that the target has informed the source that the message was rejected,
and the source has taken the necessary action. The delivery SHOULD NOT ever spontaneously attain the rejected
state at the source
```

As usual the documented behaviors are very.. unclear. It doesn't really make sense to me that we would accept the message in this case, since the message hasn't actually been received and fully processed.

**Comment by noodlefrenzy** [comment]
It seems like Released or Modified/delivery-failed would be the better options. But it _also_ seems like we might be making it too smart at that point, as we'd only want to take those actions if auto-settled is false. Anyway, I'll write my integration test and see if I can repro the behavior. I worry if that's the case, because a node failing before it can settle should not cause the message to be permanently lost.

**Comment by Romanx** [comment]
@noodlefrenzy i'm not entirely sure that was the case although it seemed to be what occurred. Reproducing it if that's the case should be simple.

I've just noticed a stupid error with my code which could have been the cause of the problem, i was calling `this.accept` with the message body as opposed to the whole message. derp. That's definitely not going to help matters.

In this case I'm not entirely sure what's happening as if I'm supposed to manually be settling the message and not doing anything, what would happen by default?

**Comment by mbroadst** [comment]
we should _definitely_ print out at least a warning message if you don't pass a valid message to one of the disposition methods. I'll pop that in, that could be very difficult to track down :)

**Comment by mbroadst** [comment]
The question now becomes: should we simply print a warning in this situation, or in fact _throw an error_. In which case it becomes very relevant for us to decide what to do when errors are thrown in the receive callback..

**Comment by noodlefrenzy** [comment]
So, I'm running the following integration test (I know, not the most elegant):

```
    it('should be able to re-receive if not acking', function(done) {
      var msgVal = uuid.v4();
      expect(config.serviceBusHost, 'Required env vars not found in ' + Object.keys(process.env)).to.exist;
      test.client = new AMQPClient(Policy.merge({
        receiverLink: {
          attach: {
            receiverSettleMode: constants.receiverSettleMode.settleOnDisposition
          }
        }
      }, Policy.ServiceBusQueue));
      var failedOnce = false;
      var timeout;
      test.client.connect(config.address)
        .then(function() {
          return Promise.all([
            test.client.createReceiver(config.defaultLink),
            test.client.createSender(config.defaultLink)
          ]);
        })
        .spread(function(receiver, sender) {
          receiver.on('message', function(message) {
            expect(message).to.exist;
            expect(message.body).to.exist;
            // Ignore messages that aren't from us.
            if (!!message.body.DataValue && message.body.DataValue === msgVal) {
              if (!failedOnce) {
                timeout = Date.now();
                failedOnce = true;
              } else {
                console.log('Took ' + (Date.now() - timeout) + ' to re-deliver');
                done();
              }
            }
          });

          return sender.send({"DataString": "From Node v2", "DataValue": msgVal});
        });
    });
```

And after ~30 seconds I see the message get re-delivered. So clearly "failing" by not ACKing the message _does_ cause it to get re-delivered as we would expect, at least for my case against my SB queue.

Other than the error you're throwing, how does your code differ from mine? And if you just avoid acking and wait 30+ seconds, do you not see a re-delivery?

**Comment by mbroadst** [comment]
frankly I'm not sure what happens if you currently throw an error inside that callback, maybe the link is corrupted somehow?

**Comment by Romanx** [comment]
I'm not calling `done();` i'm instead using `this.accept(message);` Here's a slimmed down example of what i'm doing
```javascript
var count = 0

client
  .connect(uri)
  .then(function() {
    console.log('Connected to Das Cloud');
    var queueName = topicName + '/Subscriptions/' + subscriptionName
    client.createReceiver(queueName, {         
        attach: {
            receiverSettleMode: Constants.receiverSettleMode.settleOnDisposition
        },
        creditQuantum: 1          
      })
      .then(function(receiver) {
        receiver.on('errorReceived', function(rx_err) {
            console.warn('===> RX ERROR: ', rx_err);
        });
        receiver.on('message', function(message) {     
            //Perform Actions         
            if(count > 2) throw new Error('Nooooo Error!');
            count++;
            
            //Perform More Actions
            this.accept(message);
        });
    })
});
```

**Comment by mbroadst** [comment]
@Romanx @noodlefrenzy ping on this issue
